libs/wingc/src/parser.rs

```diff
@@ -163,41 +163,74 @@ pub fn parse_wing_project(
 	// Update our collections of trees and ASTs and our file graph
 	tree_sitter_trees.insert(init_path.to_owned(), tree_sitter_tree);
 	asts.insert(init_path.to_owned(), ast);
-	file_graph.update_file(init_path, &dependent_wing_files);
+	file_graph.update_file(init_path, &dependent_wing_paths);
 
 	// Track which files still need parsing
-	let mut unparsed_files = dependent_wing_files;
+	let mut unparsed_files = dependent_wing_paths;
 
 	// Parse all remaining files in the project
-	while let Some(file_path) = unparsed_files.pop() {
+	while let Some(file_or_dir_path) = unparsed_files.pop() {
 		// Skip files that we have already seen before (they should already be parsed)
-		if files.contains_file(&file_path) {
+		if files.contains_file(&file_or_dir_path) {
 			assert!(
-				tree_sitter_trees.contains_key(&file_path),
+				tree_sitter_trees.contains_key(&file_or_dir_path),
 				"files is not in sync with tree_sitter_trees"
 			);
-			assert!(asts.contains_key(&file_path), "files is not in sync with asts");
+			assert!(asts.contains_key(&file_or_dir_path), "files is not in sync with asts");
 			assert!(
-				file_graph.contains_file(&file_path),
+				file_graph.contains_file(&file_or_dir_path),
 				"files is not in sync with file_graph"
 			);
 			continue;
 		}
 
-		let file_text = fs::read_to_string(&file_path).unwrap_or(String::new());
-		files.add_file(&file_path, file_text.clone()).unwrap();
-		files.get_file(&file_path).unwrap();
+		// Handle directories specially
+		if file_or_dir_path.is_dir() {
+			// Collect a list of all files and subdirectories in the directory
+			let mut files_and_dirs = Vec::new();
+			for entry in fs::read_dir(&file_or_dir_path).expect("read_dir call failed") {
+				let entry = entry.unwrap();
+				let path = Utf8PathBuf::from_path_buf(entry.path()).expect("invalid utf8 path");
+				if path.is_dir() || path.extension() == Some("w") {
+					files_and_dirs.push(path);
+				}
+			}
+
+			// Sort the files and directories so that we always visit them in the same order
+			files_and_dirs.sort();
+
+			// Parse the file here (since it doesn't exist)
+			let mut tree_sitter_parser = tree_sitter::Parser::new();
+			tree_sitter_parser.set_language(tree_sitter_wing::language()).unwrap();
+			let tree_sitter_tree = tree_sitter_parser.parse("", None).unwrap();
+			let ast = Scope::empty();
+			let dependent_wing_paths = files_and_dirs;
+
+			// Update our collections of trees and ASTs and our file graph
+			tree_sitter_trees.insert(file_or_dir_path.clone(), tree_sitter_tree);
+			asts.insert(file_or_dir_path.clone(), ast);
+			file_graph.update_file(&file_or_dir_path, &dependent_wing_paths);
```

Comment:
DRY with lines 228-230

Reply:
OK - I tried refactoring it to a function: [`40a12c6` (#4210)](https://github.com/winglang/wing/pull/4210/commits/40a12c63d92d34c349157b716eb5ba4e489f8af9)

I'm not really sure this improves readability, but I might be misunderstanding your suggestion

Reply:
I would have declared this as a local function because there's a lot of local state involved.

Reply:
Oh I see. Looks like it's a little tricky to do here because of borrow checking details but I'll keep the suggestion in mind!